pkg/parse/namespace.go

```diff
@@ -161,13 +162,14 @@ func (p *namespace) setSourceStatusWithRetries(ctx context.Context, newStatus so
 	currentRS := rs.DeepCopy()
 
 	setSourceStatusFields(&rs.Status.Source, p, newStatus, denominator)
+	errorSources, errorSummary := summarizeErrors(rs.Status.Status, newStatus.commit)
+	syncErrorCount := totalErrors(errorSummary)
 
-	continueSyncing := (rs.Status.Source.ErrorSummary.TotalCount == 0)
-	var errorSource []v1beta1.ErrorSource
-	if len(rs.Status.Source.Errors) > 0 {
-		errorSource = []v1beta1.ErrorSource{v1beta1.SourceError}
-	}
-	reposync.SetSyncing(&rs, continueSyncing, "Source", "Source", newStatus.commit, errorSource, rs.Status.Source.ErrorSummary, newStatus.lastUpdate)
+	// Syncing is stopped/prevented if there are any errors from the same commit.
+	// TODO: Handle non-blocking errors
+	syncing := (syncErrorCount == 0)
```

Comment:
Previously, `syncing` only depends on the source error. Now, it is changed to include all errors, including the former render or sync errors with the same commit.

It affects how the notification sends out alerts.

The current proposal is to leverage the Syncing condition to determine the sync status and send out alerts.

If the Syncing condition is `false`, it indicates the current apply loop is done. If there're no errors, then we send out `success` notification. Otherwise, it is `failed`.

In this case, if there're errors from the same commit, even if it might be fixed with the new loop, there will be a failure notification.

We can include previous errors with the same commit in the ErrorSummary to surface the errors earlier, but can we keep `syncing` only depend on the errors from the current phase (source, render, or sync)?

@sdowell FYI

Reply:
> We can include previous errors with the same commit in the ErrorSummary to surface the errors earlier, but can we keep syncing only depend on the errors from the current phase (source, render, or sync)?

It's not super clear what the Syncing condition was/is supposed to represent. 

Internally, the phases are effectively Fetch/Read -> Render/Read -> Parse -> Update (Validate/Apply/Watch/Reconcile). 

It's confusing to have the `status.sync` field represent just the errors from the Updater (which wraps the Applier) and then have the Syncing conditions which is set to True by the previous source and rendering phases.

This PR effectively assumes that since the source and rendering phases where setting the Syncing condition to True that the Syncing condition must include the entire pipeline, and not just the Update/Apply phase.

If that is NOT what we want, then we need to document the proposed change and implement it consistency.

Reply:
Also, it's worth noting that the inputs to a sync/apply are not just Git. They also can include Helm values in the spec and other config values, like Git dir, URL, cluster name, etc. So it doesn't make sense to have a unique pipeline execution depend on just a commit. Even just a generation would be technically insufficient, because the previous run may fail, perhaps due to some ephemeral network issue, triggering a retry. Today we don't indicate what generation or attempt/retry the Syncing condition represents, which makes it hard to use for triggering discrete alerts.

Reply:
The Syncing condition represents the _**current/latest**_ fetch -> render -> parse -> update pipeline, while `status.source`, `status.rendering`, and `status.sync` represent the statuses of the last processed commit in each phase.

It is possible that the Syncing condition only includes status of **_partial_** pipeline while the loop is still in progress.

For example,
1. The first pipeline failed at commit A in the apply phase, `status.sync` will have errors for commit A.
2. The second pipeline failed to fetch commit A, `status.source` will have errors for commit A, and `status.sync` still contains the former apply errors.
3. The current pipeline succeeded in fetching and rendering commit A, source errors from the second pipeline will get cleared from `status.source`. Then the `parseAndUpdate` function will call `p.setSourceStatus`, which calls the `setSourceStatusWithRetries` function. The `setSourceStatusWithRetries` function should set Syncing condition to `true` because it will continue applying the parsed objects next. However, this PR sets the syncing condition to `false` because of the sync errors from the first pipeline.